drivers/timer/nrf_grtc_timer.c

```diff
@@ -551,22 +553,47 @@ void sys_clock_set_timeout(int32_t ticks, bool idle)
 {
 	ARG_UNUSED(idle);
 
+	if (ticks == 0) {
+		return;
+	}
+
 	if (!IS_ENABLED(CONFIG_TICKLESS_KERNEL)) {
 		return;
 	}
 
-	ticks = (ticks == K_TICKS_FOREVER) ? MAX_TICKS : MIN(MAX_TICKS, MAX(ticks, 0));
+	uint32_t ch = system_clock_channel_data.channel;
 
-	uint64_t delta_time = ticks * CYC_PER_TICK;
+	if ((cc_value == expired_cc) && (ticks < MAX_REL_TICKS)) {
```

Comment:
FWIW, for the specific case you're trying to handle here (re-setting the timeout at the end of the handler), you can detect that even faster by seeing that you're being called inside the timer ISR.  Set a static boolean before calling the handler and clear it after it returns.  That would spend the two extra memory operations for every ISR, but would avoid having to call cc_value_read() on every call to set_timeout, which is almost certainly a win.

Reply:
I had that approach initially, but it was failing in a stress test which is calling calling k_timer from interrupt context which has higher priority than the system timer interrupt and I had to change it to this condition so only first attempt that follows timeout expiration is using optimize setting.